gdal-3.2.1/man/man1/gdal_pansharpen.1
.\" Man page generated from reStructuredText.
.
.TH "GDAL_PANSHARPEN" "1" "Dec 29, 2020" "" "GDAL"
.SH NAME
gdal_pansharpen \-  Perform a pansharpen operation.
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen [\-\-help\-general] pan_dataset {spectral_dataset[,band=num]}+ out_dataset
                [\-of format] [\-b band]* [\-w weight_val]*
                [\-r {nearest,bilinear,cubic,cubicspline,lanczos,average}]
                [\-threads {ALL_CPUS|number}] [\-bitdepth val] [\-nodata val]
                [\-spat_adjust {union,intersection,none,nonewithoutwarning}]
                [\-co NAME=VALUE]* [\-q]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
The \fBgdal_pansharpen.py\fP script performs a pan\-sharpening operation. It
can create a "classic" output dataset (such as GeoTIFF), or a VRT
dataset describing the pan\-sharpening operation.
.sp
More details can be found in the gdal_vrttut_pansharpen section.
.INDENT 0.0
.TP
.B \-of <format>:
Select the output format. Starting with GDAL 2.3, if not specified,
the format is guessed from the extension (previously was \fBGTiff\fP). Use
the short format name. \fBVRT\fP can also be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-b <band>
Select band \fIband\fP from the input spectral bands for output. Bands
are numbered from 1 in the order spectral bands are specified.
Multiple \fB\-b\fP switches may be used. When no \-b switch is used, all
input spectral bands are set for output.
.UNINDENT
.INDENT 0.0
.TP
.B \-w <weight_val>
Specify a weight for the computation of the pseudo panchromatic
value. There must be as many \-w switches as input spectral bands.
.UNINDENT
.INDENT 0.0
.TP
.B \-r {nearest,bilinear,cubic (default),cubicspline,lanczos,average}
Select a resampling algorithm.
.UNINDENT
.INDENT 0.0
.TP
.B \-threads {ALL_CPUS,number}
Specify number of threads to use to do the resampling and
pan\-sharpening itself. Can be an integer number or ALL_CPUS.
.UNINDENT
.INDENT 0.0
.TP
.B \-bitdepth <val>
Specify the bit depth of the panchromatic and spectral bands (e.g.
12). If not specified, the NBITS metadata item from the panchromatic
band will be used if it exists.
.UNINDENT
.INDENT 0.0
.TP
.B \-nodata <val>
Specify nodata value for bands. Used for the resampling and
pan\-sharpening computation itself. If not set, deduced from the
input bands, provided they have a consistent setting.
.UNINDENT
.INDENT 0.0
.TP
.B \-spat_adjust {union(default),intersection,none,nonewithoutwarning}
Select behavior when bands have not the same extent. See
\fISpatialExtentAdjustment\fP documentation in gdal_vrttut_pansharpen
.UNINDENT
.INDENT 0.0
.TP
.B \-co <NAME=VALUE>
Many formats have one or more optional creation options that can be
used to control particulars about the file created. For instance,
the GeoTIFF driver supports creation options to control compression,
and whether the file should be tiled.
.sp
The creation options available vary by format driver, and some
simple formats have no creation options at all. A list of options
supported for a format can be listed with the
\-\-formats
command line option but the documentation for the format is the
definitive source of information on driver creation options.
See raster_drivers format
specific documentation for legal creation options for each format.
.UNINDENT
.INDENT 0.0
.TP
.B \-q
Suppress progress monitor and other non\-error output.
.UNINDENT
.INDENT 0.0
.TP
.B <pan_dataset>
Dataset with panchromatic band (first band will be used).
.UNINDENT
.INDENT 0.0
.TP
.B <spectral_dataset>[,band=num]
Dataset with one or several spectral bands. If the band option is
not specified, all bands of the datasets are taken into account.
Otherwise, only the specified (num)th band. The same dataset can be
repeated several times.
.UNINDENT
.INDENT 0.0
.TP
.B <out_dataset>
Output dataset
.UNINDENT
.sp
Bands should be in the same projection.
.SH EXAMPLE
.sp
With spectral bands in a single dataset :
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen.py panchro.tif multispectral.tif pansharpened_out.tif
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
With a few spectral bands from a single dataset, reordered :
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen.py panchro.tif multispectral.tif,band=3 multispectral.tif,band=2 multispectral.tif,band=1 pansharpened_out.tif
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
With spectral bands in several datasets :
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen.py panchro.tif band1.tif band2.tif band3.tif pansharpened_out.tif
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Specify weights:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen.py \-w 0.7 \-w 0.2 \-w 0.1 panchro.tif multispectral.tif pansharpened_out.tif
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Specify RGB bands from a RGBNir multispectral dataset while computing
the pseudo panchromatic intensity on the 4 RGBNir bands:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
gdal_pansharpen.py \-b 1 \-b 2 \-b 3 panchro.tif rgbnir.tif pansharpened_out.tif
.ft P
.fi
.UNINDENT
.UNINDENT
.SH AUTHOR
Even Rouault <even.rouault@spatialys.com>
.SH COPYRIGHT
1998-2020
.\" Generated by docutils manpage writer.
.


gdal-3.4.0/swig/python/gdal-utils/osgeo_utils/gdal_pansharpen.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
###############################################################################
# $Id: gdal_pansharpen.py 1481a885e2de717aa8834045768a07548d7a1fb3 2021-10-27 21:11:07 +0200 Even Rouault $
#
#  Project:  GDAL scripts
#  Purpose:  Perform a pansharpening operation
#  Author:   Even Rouault <even.rouault at spatialys.com>
#
###############################################################################
#  Copyright (c) 2015, Even Rouault <even.rouault at spatialys.com>
#  Copyright (c) 2021, Idan Miara <idan@miara.com>
#
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included
#  in all copies or substantial portions of the Software.
#
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
#  OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.
###############################################################################

import os
import os.path
import sys
from numbers import Real
from typing import Optional, Sequence, Union, List

from osgeo import gdal
from osgeo_utils.auxiliary.util import GetOutputDriverFor


def Usage():
    print('Usage: gdal_pansharpen [--help-general] pan_dataset {spectral_dataset[,band=num]}+ out_dataset')
    print('                       [-of format] [-b band]* [-w weight]*')
    print('                       [-r {nearest,bilinear,cubic,cubicspline,lanczos,average}]')
    print('                       [-threads {ALL_CPUS|number}] [-bitdepth val] [-nodata val]')
    print('                       [-spat_adjust {union,intersection,none,nonewithoutwarning}]')
    print('                       [-verbose_vrt] [-co NAME=VALUE]* [-q]')
    print('')
    print('Create a dataset resulting from a pansharpening operation.')
    return -1


def main(argv):
    argv = gdal.GeneralCmdLineProcessor(argv)
    if argv is None:
        return -1

    pan_name = None
    spectral_names = []
    spectral_ds = []
    spectral_bands = []
    band_nums = []
    weights = []
    driver_name = None
    creation_options = []
    progress_callback = gdal.TermProgress_nocb
    resampling = None
    spat_adjust = None
    verbose_vrt = False
    num_threads = None
    bitdepth = None
    nodata_value = None

    i = 1
    argc = len(argv)
    while i < argc:
        if (argv[i] == '-of' or argv[i] == '-f') and i < len(argv) - 1:
            driver_name = argv[i + 1]
            i = i + 1
        elif argv[i] == '-r' and i < len(argv) - 1:
            resampling = argv[i + 1]
            i = i + 1
        elif argv[i] == '-spat_adjust' and i < len(argv) - 1:
            spat_adjust = argv[i + 1]
            i = i + 1
        elif argv[i] == '-b' and i < len(argv) - 1:
            band_nums.append(int(argv[i + 1]))
            i = i + 1
        elif argv[i] == '-w' and i < len(argv) - 1:
            weights.append(float(argv[i + 1]))
            i = i + 1
        elif argv[i] == '-co' and i < len(argv) - 1:
            creation_options.append(argv[i + 1])
            i = i + 1
        elif argv[i] == '-threads' and i < len(argv) - 1:
            num_threads = argv[i + 1]
            i = i + 1
        elif argv[i] == '-bitdepth' and i < len(argv) - 1:
            bitdepth = argv[i + 1]
            i = i + 1
        elif argv[i] == '-nodata' and i < len(argv) - 1:
            nodata_value = argv[i + 1]
            i = i + 1
        elif argv[i] == '-q':
            progress_callback = None
        elif argv[i] == '-verbose_vrt':
            verbose_vrt = True
        elif argv[i][0] == '-':
            sys.stderr.write('Unrecognized option : %s\n' % argv[i])
            return Usage()
        elif pan_name is None:
            pan_name = argv[i]
        else:
            spectral_names.append(argv[i])

        i = i + 1

    if pan_name is None or len(spectral_names) < 2:
        return Usage()

    dst_filename = spectral_names.pop()
    return gdal_pansharpen(argv=None,
                           pan_name=pan_name,
                           spectral_names=spectral_names, spectral_ds=spectral_ds, spectral_bands=spectral_bands,
                           band_nums=band_nums, weights=weights,
                           dst_filename=dst_filename, driver_name=driver_name, creation_options=creation_options,
                           resampling=resampling, spat_adjust=spat_adjust,
                           num_threads=num_threads, bitdepth=bitdepth, nodata_value=nodata_value,
                           verbose_vrt=verbose_vrt,
                           progress_callback=progress_callback)


def gdal_pansharpen(argv: Optional[Sequence[str]] = None,
                    pan_name: Optional[str] = None,
                    spectral_names: Optional[Sequence[str]] = None,
                    spectral_ds: Optional[List[gdal.Dataset]] = None, spectral_bands: Optional[List[gdal.Band]] = None,
                    band_nums: Optional[Sequence[int]] = None, weights: Optional[Sequence[float]] = None,
                    dst_filename: Optional[str] = None, driver_name: Optional[str] = None,
                    creation_options: Optional[Sequence[str]] = None,
                    resampling: Optional[str] = None, spat_adjust: Optional[str] = None,
                    num_threads: Optional[Union[int, str]] = None, bitdepth: Optional[Union[int, str]] = None,
                    nodata_value: Optional[Union[Real, str]] = None,
                    verbose_vrt: bool = False,
                    progress_callback: Optional = gdal.TermProgress_nocb):
    if argv:
        # this is here for backwards compatibility
        return main(argv)

    spectral_names = spectral_names or []
    spectral_ds = spectral_ds or []
    spectral_bands = spectral_bands or []
    band_nums = band_nums or []
    weights = weights or []
    creation_options = creation_options or []

    if spectral_names:
        parse_spectral_names(spectral_names=spectral_names, spectral_ds=spectral_ds, spectral_bands=spectral_bands)

    if pan_name is None or not spectral_bands:
        return 1

    pan_ds = gdal.Open(pan_name)
    if pan_ds is None:
        return 1

    if driver_name is None:
        driver_name = GetOutputDriverFor(dst_filename)

    if not band_nums:
        band_nums = [j + 1 for j in range(len(spectral_bands))]
    else:
        for band in band_nums:
            if band < 0 or band > len(spectral_bands):
                print('Invalid band number in -b: %d' % band)
                return 1

    if weights and len(weights) != len(spectral_bands):
        print('There must be as many -w values specified as input spectral bands')
        return 1

    vrt_xml = """<VRTDataset subClass="VRTPansharpenedDataset">\n"""
    if band_nums != [j + 1 for j in range(len(spectral_bands))]:
        for i, band in enumerate(band_nums):
            sband = spectral_bands[band - 1]
            datatype = gdal.GetDataTypeName(sband.DataType)
            colorname = gdal.GetColorInterpretationName(sband.GetColorInterpretation())
            vrt_xml += """  <VRTRasterBand dataType="%s" band="%d" subClass="VRTPansharpenedRasterBand">
      <ColorInterp>%s</ColorInterp>
  </VRTRasterBand>\n""" % (datatype, i + 1, colorname)

    vrt_xml += """  <PansharpeningOptions>\n"""

    if weights:
        vrt_xml += """      <AlgorithmOptions>\n"""
        vrt_xml += """        <Weights>"""
        for i, weight in enumerate(weights):
            if i > 0:
                vrt_xml += ","
            vrt_xml += "%.16g" % weight
        vrt_xml += "</Weights>\n"
        vrt_xml += """      </AlgorithmOptions>\n"""

    if resampling is not None:
        vrt_xml += f'      <Resampling>{resampling}</Resampling>\n'

    if num_threads is not None:
        vrt_xml += f'      <NumThreads>{num_threads}</NumThreads>\n'

    if bitdepth is not None:
        vrt_xml += f'      <BitDepth>{bitdepth}</BitDepth>\n'

    if nodata_value is not None:
        vrt_xml += f'      <NoData>{nodata_value}</NoData>\n'

    if spat_adjust is not None:
        vrt_xml += f'      <SpatialExtentAdjustment>{spat_adjust}</SpatialExtentAdjustment>\n'

    pan_relative = '0'
    if driver_name.upper() == 'VRT':
        if not os.path.isabs(pan_name):
            pan_relative = '1'
            pan_name = os.path.relpath(pan_name, os.path.dirname(dst_filename))

    vrt_xml += """    <PanchroBand>
      <SourceFilename relativeToVRT="%s">%s</SourceFilename>
      <SourceBand>1</SourceBand>
    </PanchroBand>\n""" % (pan_relative, pan_name)

    for i, sband in enumerate(spectral_bands):
        dstband = ''
        for j, band in enumerate(band_nums):
            if i + 1 == band:
                dstband = ' dstBand="%d"' % (j + 1)
                break

        ms_relative = '0'
        ms_name = spectral_ds[i].GetDescription()
        if driver_name.upper() == 'VRT':
            if not os.path.isabs(ms_name):
                ms_relative = '1'
                ms_name = os.path.relpath(ms_name, os.path.dirname(dst_filename))

        vrt_xml += """    <SpectralBand%s>
      <SourceFilename relativeToVRT="%s">%s</SourceFilename>
      <SourceBand>%d</SourceBand>
    </SpectralBand>\n""" % (dstband, ms_relative, ms_name, sband.GetBand())

    vrt_xml += """  </PansharpeningOptions>\n"""
    vrt_xml += """</VRTDataset>\n"""

    if driver_name.upper() == 'VRT':
        f = gdal.VSIFOpenL(dst_filename, 'wb')
        if f is None:
            print('Cannot create %s' % dst_filename)
            return 1
        gdal.VSIFWriteL(vrt_xml, 1, len(vrt_xml), f)
        gdal.VSIFCloseL(f)
        if verbose_vrt:
            vrt_ds = gdal.Open(dst_filename, gdal.GA_Update)
            vrt_ds.SetMetadata(vrt_ds.GetMetadata())
        else:
            vrt_ds = gdal.Open(dst_filename)
        if vrt_ds is None:
            return 1

        return 0

    vrt_ds = gdal.Open(vrt_xml)
    out_ds = gdal.GetDriverByName(driver_name).CreateCopy(dst_filename, vrt_ds, 0, creation_options,
                                                          callback=progress_callback)
    if out_ds is None:
        return 1
    return 0


def parse_spectral_names(spectral_names: Sequence[str],
                         spectral_ds: List[gdal.Dataset],
                         spectral_bands: List[gdal.Band]):
    for spectral_arg in spectral_names:
        # add selected bands
        pos = spectral_arg.find(',band=')
        if pos > 0:
            spectral_name = spectral_arg[0:pos]
            ds = gdal.Open(spectral_name)
            if ds is None:
                return 1
            band_num = int(spectral_arg[pos + len(',band='):])
            band = ds.GetRasterBand(band_num)
            spectral_ds.append(ds)
            spectral_bands.append(band)
        else:
            # add all bands
            spectral_name = spectral_arg
            ds = gdal.Open(spectral_name)
            if ds is None:
                return 1
            for j in range(ds.RasterCount):
                spectral_ds.append(ds)
                spectral_bands.append(ds.GetRasterBand(j + 1))


if __name__ == '__main__':
    sys.exit(main(sys.argv))
